collect_app/src/test/java/org/odk/collect/android/widgets/TriggerWidgetTest.java

```diff
@@ -1,67 +1,98 @@
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;
+
+@RunWith(RobolectricTestRunner.class)
+public class TriggerWidgetTest {
+
+    @Test
+    public void getAnswer_whenPromptAnswerDoesNotHaveAnswer_returnsNull() {
+        assertThat(createWidget(promptWithAnswer(null)).getAnswer(), nullValue());
+    }
+
+    @Test
+    public void getAnswer_whenPromptHasAnswer_returnsAnswer() {
+        TriggerWidget widget = createWidget(promptWithAnswer(new StringData("OK")));
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("OK"));
+    }
+
+    @Test
+    public void clearAnswer_clearsWidgetAnswer() {
+        TriggerWidget widget = createWidget(promptWithAnswer(new StringData("OK")));
+
+        widget.clearAnswer();
+        assertThat(widget.getAnswer(), nullValue());
     }
 
-    @NonNull
-    @Override
-    public StringData getNextAnswer() {
-        return new StringData(TriggerWidget.OK_TEXT);
+    @Test
+    public void clearAnswer_callsValueChangeListeners() {
+        TriggerWidget widget = createWidget(promptWithAnswer(null));
+        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
+
+        widget.clearAnswer();
+        verify(valueChangedListener).widgetValueChanged(widget);
     }
 
-    @Override
-    public void getAnswerShouldReturnExistingAnswerIfPromptHasExistingAnswer() {
-        super.getAnswerShouldReturnExistingAnswerIfPromptHasExistingAnswer();
-        assertTrue(getWidget().getTriggerButton().isChecked());
+    @Test
+    public void usingReadOnlyOption_makesAllClickableElementsDisabled() {
+        TriggerWidget widget = createWidget(promptWithReadOnly());
+        assertThat(widget.getCheckBox().getVisibility(), equalTo(View.VISIBLE));
+        assertThat(widget.getCheckBox().isEnabled(), equalTo(Boolean.FALSE));
     }
 
     @Test
-    public void checkingTheTriggerBoxShouldSetTheAnswer() {
-        TriggerWidget widget = getWidget();
-        assertNull(widget.getAnswer());
+    public void whenPromptAnswerDoesNotHaveAnswer_checkboxIsUnchecked() {
+        TriggerWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(widget.getCheckBox().isChecked(), equalTo(false));
+    }
 
-        CheckBox triggerButton = widget.getTriggerButton();
-        assertFalse(triggerButton.isChecked());
+    @Test
+    public void whenPromptHasAnswer_checkboxIsChecked() {
+        TriggerWidget widget = createWidget(promptWithAnswer(new StringData("OK")));
+        assertThat(widget.getCheckBox().isChecked(), equalTo(true));
+    }
+
+    @Test
+    public void checkingCheckbox_setsAnswer() {
+        TriggerWidget widget = createWidget(promptWithAnswer(null));
+        CheckBox triggerButton = widget.getCheckBox();
 
         triggerButton.setChecked(true);
-        triggerButton.callOnClick();
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("OK"));
```

Comment:
This makes me so angry. Why "OK"?! Everything about this trigger widget is objectionable starting from the fact that `trigger` is a [well-defined concept in W3C XForms](https://en.wikibooks.org/wiki/XForms/Trigger) and that this is something else entirely.

No action to take, just venting. 😾

Reply:
🔥